lib/online_order.rb

```diff
@@ -0,0 +1,85 @@
+require 'csv'
+require_relative 'order'
+require_relative 'customer'
+
+ONLINE_FILE = 'support/online_orders.csv'
+
+module Grocery
+  class OnlineOrder < Order
+    attr_reader :status, :customer
+
+    def initialize(id, products, customer_id, status=:pending)
+      super(id, products)
+      @customer = Customer.find(customer_id)
+      @status = status
+    end
+
+    # if the online order instance has no products, returns the total provided by Order#total (no shipping)
+    # otherwise, returns total provided by Order#total + $10 for shipping
+    def total
+      super
+      if @products.empty?
+        return super
+      else
+        return super + 10
+      end
+    end
+
+    # returns false if the product name is already included in the online order's products,
+    # raises an ArgumentError if the online order status is processing, shipped or complete,
+    # otherwise adds product to the instance's product hash and returns true
+    def add_product(product_name, product_price)
+      if @products.keys.include?(product_name)
+        return false
+      elsif [:processing, :shipped, :complete].include?(@status)
```

Comment:
Most of the logic for this method already exists in `Order#add_product`. You should be able to use `super` to DRY things up - something like:

```ruby
def add_product(product_name, product_price)
  if [:processing, :shipped, :complete].include?(@status)
    raise ArgumentError.new("The order status is #{@status}. New products can no longer be added to this order.")
  end
  super(product_name, product_price)
end
```